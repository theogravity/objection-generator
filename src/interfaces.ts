export interface YamlConfig {
  config: {
    model: {
      classNamePrefix?: string
      classNamePostfix?: string
    }
  }
  models: {
    [key: string]: ModelDef
  }
  components: {
    [key: string]: any
  }
}

export interface ModelDef {
  tableName: string
  jsonSchema: {
    required?: string[]
    properties: {
      [key: string]: {
        type: string
        [key: string]: any
      }
    }
  }
}

export interface Arguments {
  specFile: string
  outDir: string
  templateDir?: string
}
