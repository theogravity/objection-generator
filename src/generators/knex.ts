import Handlebars from 'handlebars'
import { promises } from 'fs'
import fsEx from 'fs-extra'

import { BaseGenerator } from './base'

export class KnexGenerator extends BaseGenerator {
  async generate () {
    await fsEx.ensureDir(this.outDir + '/migrations')
    await this.generateMigration()
    fsEx.copy(`${this.templatesDir}/migrate.js`, `${this.outDir}/migrate.js`)
  }

  protected async generateMigration () {
    for (const file of this.templateFiles) {
      if (file.includes('000-init.hbs')) {
        const hbsData = await promises.readFile(file, 'utf8')
        const template = Handlebars.compile(hbsData)

        const templateData = {
          models: this.defs.models
        }

        const content = template(templateData)
        await promises.writeFile(
          `${this.outDir}/migrations/000-init.js`,
          this.prettify(content),
          'utf8'
        )
      }
    }
  }
}
