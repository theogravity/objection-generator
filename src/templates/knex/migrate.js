const { resolve } = require('path')
const kn = require('knex')

;(async () => {
  // Sample migration with a sqlite database
  const knex = kn({
    client: 'sqlite3',
    connection: {
      filename: resolve(process.cwd(), 'sample.db')
    },
    migrations: {
      tableName: 'knex_migrations',
      directory: resolve(__dirname, 'migrations')
    },
    timezone: 'UTC'
  })

  // run the migration
  await knex.migrate.latest()

  console.log(`Migration complete. Created 'sample.db'`)
  process.exit(0)
})()
