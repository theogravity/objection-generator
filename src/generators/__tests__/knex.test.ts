/* eslint-env jest */

import { pathExistsSync, remove } from 'fs-extra'
import { readFileSync } from 'fs'
import { resolve } from 'path'
import validConfig from '../../__fixtures__/valid-config.json'

import { KnexGenerator } from '../knex'
import { YamlConfig } from '../../interfaces'

const templateDir = resolve(__dirname, '..', '..', 'templates')

const OUT_DIR = '/tmp/objection-generator-test/generator/knex'

beforeEach(async () => {
  await remove(OUT_DIR)
})

describe('KnexGenerator class', () => {
  it('should generate a migration', async () => {
    const generator = new KnexGenerator(
      (validConfig as unknown) as YamlConfig,
      templateDir + '/knex',
      OUT_DIR
    )

    await generator.init()
    await generator.generate()

    const migrationPath = resolve(OUT_DIR, 'migrations', '000-init.js')

    expect(pathExistsSync(migrationPath)).toBe(true)
    expect(readFileSync(migrationPath, 'utf8')).toMatchSnapshot()
  })
})
