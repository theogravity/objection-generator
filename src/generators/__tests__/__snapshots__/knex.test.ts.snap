// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KnexGenerator class should generate a migration 1`] = `
"async function up (knex) {
  await knex.schema.createTable('persons', table => {
    table.string('id')
    table.string('name', 100).notNullable()
    table.integer('age')
    table.enu('gender', ['Male', 'Female', 'Other']).defaultTo('Female')
    table.string('username', 25).notNullable()

    table.primary(['id'])
  })
  await knex.schema.createTable('movies', table => {
    table.string('id')
    table.string('name', 255).notNullable()

    table.primary(['id'])
  })
  await knex.schema.createTable('reviews', table => {
    table.string('review_id')
    table.string('author_id').notNullable()
    table.string('movie_id').notNullable()
    table.string('content')

    table.primary(['review_id'])
  })
}

async function down (knex) {
  await knex.schema.dropTable('persons')
  await knex.schema.dropTable('movies')
  await knex.schema.dropTable('reviews')
}

module.exports = {
  up,
  down
}
"
`;
